### PR TITLE
Map Server Blob_Authorization_Failure to Router.

### LINKED ISSUE #1001: Map StoreErrorCodes.Authorization_Failure to ServerErrorCode.Blob_Authorization_Failure
In ErrorMapping and add test.

### PR DESCRIPTION
**Pull request by zzmao**
Fix #1001.

### CODE REVIEW

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 1]
```diff
@@ -352,6 +352,55 @@ void check() throws Exception {
...
+  /**
+   * Submit a getBlob operation with incorrect accountId/ContainerId in blobId..
+   * @param opChain the {@link OperationChain} object that this operation is a part of.
+   */
+  private void startGetBlobAuthorizationFailTest(final OperationChain opChain) {
+    Callback<GetBlobResult> callback = new TestCallback<>(opChain, false);
+    BlobId originalId, fraudId = null;
+    try {
+      originalId = new BlobId(opChain.blobId, clusterMap);
+      fraudId = BlobId.craft(originalId, originalId.getVersion(), (short) (1), (short) 1);
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: what if the account id, container id is 1? I think its probably better to do something like "current id + 1" so you are sure it is different

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 2]
```diff
@@ -352,6 +352,55 @@ void check() throws Exception {
...
+    } catch (IOException e) {
+      Assert.fail("BlobId creation failed.");
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: just let the exception through. Provides more info than this `fail()`

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 1 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: Same with the delete

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 2 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: Same with the delete

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 3]
```diff
@@ -591,7 +662,14 @@ public void onCompletion(T result, Exception exception) {
         return;
       }
       action(result);
-      continueChain(opChain);
+      try {
+        continueChain(opChain);
+      } catch (IOException e) {
+        if (expectError) {
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: ! (not)

Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: This way to throw `IOException` does't work, it just terminated the test earlier. Test passed if I throw an `IOException`. 
It's `TestFuture.check()` do the check, so I tried another way which throw `IOException` in `check()`

Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: The new way can throw sth like this;

```
java.io.IOException: hahaha

	at com.github.ambry.server.RouterServerTestFramework.startGetBlobAuthorizationFailTest(RouterServerTestFramework.java:367)
	at com.github.ambry.server.RouterServerTestFramework.continueChain(RouterServerTestFramework.java:504)
	at com.github.ambry.server.RouterServerTestFramework.access$300(RouterServerTestFramework.java:65)
	at com.github.ambry.server.RouterServerTestFramework$TestCallback.onCompletion(RouterServerTestFramework.java:695)
	at com.github.ambry.router.NonBlockingRouter$1.onCompletion(NonBlockingRouter.java:153)
	at com.github.ambry.router.NonBlockingRouter$1.onCompletion(NonBlockingRouter.java:147)
	at com.github.ambry.router.NonBlockingRouter.completeOperation(NonBlockingRouter.java:465)
	at com.github.ambry.router.NonBlockingRouter.completeOperation(NonBlockingRouter.java:442)
	at com.github.ambry.router.GetBlobInfoOperation.checkAndMaybeComplete(GetBlobInfoOperation.java:437)
	at com.github.ambry.router.GetBlobInfoOperation.handleResponse(GetBlobInfoOperation.java:251)
	at com.github.ambry.router.GetManager.handleResponse(GetManager.java:224)
	at com.github.ambry.router.NonBlockingRouter$OperationController.onResponse(NonBlockingRouter.java:688)
	at com.github.ambry.router.NonBlockingRouter$OperationController.run(NonBlockingRouter.java:721)
	at java.lang.Thread.run(Thread.java:745)

```

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 4]
```diff
@@ -564,6 +653,21 @@ void checkDeleted() throws Exception {
       }
     }
 
+    /**
+     * Check if router got expected RouterErrorCode.
+     * @throws Exception
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: please add java doc for `routerErrorCode`.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 5]
```diff
@@ -352,6 +353,82 @@ void check() throws Exception {
     opChain.testFutures.add(testFuture);
   }
 
+  /**
+   * Submit a getBlob operation with incorrect accountId/ContainerId in blobId..
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: minor: extra period.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 6]
```diff
@@ -564,6 +653,21 @@ void checkDeleted() throws Exception {
       }
     }
 
+    /**
+     * Check if router got expected RouterErrorCode.
+     * @throws Exception
+     */
+    void checkExpectedRouterErrorCode(RouterErrorCode routerErrorCode) throws Exception {
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: it seems the `Exception` is unnecessary, we can remove it.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 7]
```diff
@@ -352,6 +353,82 @@ void check() throws Exception {
...
+      }
+    };
+    opChain.testFutures.add(testFuture);
+  }
+
+  /**
+   * Submit a deleteBlob operation with incorrect accountId/ContainerId in blobId.
+   * @param opChain the {@link OperationChain} object that this operation is a part of.
+   */
+  private void startDeleteBlobAuthorizationFailTest(final OperationChain opChain) {
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: quick question: do we have blob authorization succeeded tests for `delete` and `get` operations?

Comment (jsjtzyy) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: never mind. I notice that the normal `get` and `delete` will succeed, that's ok then.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 4 above)
Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: added.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 5 above)
Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: sure.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 8]
```diff
@@ -352,6 +353,82 @@ void check() throws Exception {
...
+    BlobId originalId, fraudId;
+    try {
+      originalId = new BlobId(opChain.blobId, clusterMap);
+      fraudId = BlobId.craft(originalId, originalId.getVersion(), (short) (originalId.getAccountId() + 1),
+          (short) (originalId.getContainerId() + 1));
+    } catch (IOException e) {
+      // If there is a blobId creation failure, throw the exception and don't need to do actual action.
+      FutureResult<Void> future = new FutureResult<>();
+      // continue the chain
+      future.done(null, null);
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: `future.done(null, e)`?

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 9]
```diff
@@ -352,6 +353,82 @@ void check() throws Exception {
...
+      callback.onCompletion(null, null);
```
Comment (vgkholla) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: `callback.onCompletion(null, e);`?

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 8 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: same in delete

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 9 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: same in delete

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 8 above)
Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: changed.

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 9 above)
Comment (zzmao) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: changed.

**Comment by vgkholla** [comment]
Will merge after travis build passes

**Pull request merged**